### ISSUE TITLE
lz4 does not remove input file after compression (incompatibility with gzip, etc.)

### ISSUE THREAD

**Issue by jimav** [lz4/lz4#170]
The "lz4" (not "lz4c") command is intended to be command-line compatible with gzip, bzip2, etc.
in many cases.  For why this is important to world sanity, please read
https://code.google.com/p/lz4/issues/detail?id=83&can=1&q=gzip

However, "lz4" is not compatible with gzip in the most basic use case:

   lz4 filename
1.  After creating filename.lz4 the input is not removed
2.  If the command is in a script with stdout redirected, compressed data is written to
    stdout and filename.lz4 is not created at all.

"unlz4" has the same compatibility bugs.

To recap what was said by others in the above-referenced bug report, while it is true that changing the options will break some existing scripts, it is far more important to quash this disturbance to the continuum now before it propagates pain and suffering for decades to come.   The expressed desire (according to the lz4 man page and comments in the above "fixed" bug) is to make "lz4" (but not "lz4c") compatible with the decades of settled practice with options to gzip, bzip2, etc.    So fixing these problems should not cause anyone much surprise.

Suggest maintaining the old non-standard/incompatible behavior when invoked as "lz4c", but make lz4 (and unlz4 and lz4cat) work as exactly like gzip etc. as practicable without undue effort.

Here are the basics of "gzip" compatible behavior:
1. Never read/write date from stdin/stdout if a file name is specifed, unless the file name 
     is "-" or an option like -c --stdout or --to-stdout is given explicitly.
   
   Whether or not stdout is a tty should never -change- the output destination
   (but it's still ok to refuse to write compressed data to a tty without -c)
2.  lz4 FILE1 [FILE2 ...]       should always compress each file and create FILE.lz4 etc.
     -and- remove the input files.   Never interpret arguments as output file names.  
    (and analogously for unlz4 or the equivalent lz4 -d FILE...)
   
    Remove the --multiple option (make it the default and only behavior)
3.  lz4 (no args)   or   lz4 -
      Silently read stdin, decompress, and write to stdout
      (and analogously for unlz4 and lz4 -d)

**Comment by Cyan4973** [comment]
> lz4 FILE1 [FILE2 ...] should always compress each file and create FILE.lz4 etc.
> (and analogously for unlz4 or the equivalent lz4 -d FILE...)
> Never interpret arguments as output file names.
> Remove the --multiple option (make it the default and only behavior)

Well, this capability has only been created recently.
Now that it's completed, it may look an easy feature to build, but back then it was not. Therefore, it wasn't implemented at the time `lz4` was created out of `lz4c` to become more compatible with `gzip` ("more compatible" never was intended to be understood as 100% identical).
Hence for many years `lz4` kept its initial design, where :
`lz4 file1 file2` means "compress file1 into file2"

As multiple-files compression capability is now present, it may seem simple to make it the default behavior now. Technically, it's simple. 

My worry is that existing users may find their scripts broken by this fairly important command line change. We are no longer in a situation where lz4 was relatively new, it's now deployed in a very large number of systems, and breaking existing user base should not be treated lightly.

This is the only reason why `--multiple` is not currently default, but it's a damn good one.

> If the command is in a script with stdout redirected, compressed data is written to stdout and filename.lz4 is not created at all.
> Never read/write date from stdin/stdout if a file name is specifed, unless the file name 
is "-" or an option like -c --stdout or --to-stdout is given explicitly.

Interesting, I wasn't aware of this gzip behavior.
What is going to be redirected to `stdout` in such a case ?
That behavior could indeed be added in association with `-m` command.

> lz4 (no args) or lz4 -
> Silently read stdin, decompress, and write to stdout

Isn't that the way `lz4` works today ?
The information update is turned off as soon as `lz4` is used in piped mode.
On top of that, I'm sure you have noticed the `-q` command which turns if off completely.

> After creating filename.lz4 the input is not removed

Well, I suspect we will not agree on this one.

Silently erasing by default the source file is an incredible responsibility outreach.
I can maybe understand it in a context where storage is expensive and scarce, as it used to be in the early 90's. And even then, I'm not sure to understand why it was designed as "silent default" instead of "explicit opt-in".

But in 2015, this is just a plain design fault. And there is no more any excuse for it : storage is plentiful, mistakes are easy, roll back capability is paramount in case of a mis-type, and `rm` is not kind for this objective.

So basically this is a permanent no. Removing a file _must_ be an explicit order.


To sum up : 
I see 2 missing capabilities : 

- compress files to *.lz4 even if redirection to stdout is enabled when `-m` option is selected
- an opt-in command to remove source files after their compression (for example `--rm`)

and then the question is about what should be default or not.

Preserving compatibility with existing user base is paramount, don't expect changes which could break an unknown amount of existing systems.

Alternatively, it's quite easy to create a new shortcut (for example `lz4z`) which simply enables a bunch of opt-in commands so that the behavior is more the one you describe, such as
`lz4z == lz4 --multiple --rm --quiet` .

I'm just wondering if this is really useful, as unix scripts I know of tend to define compressors through variables, and it's typically not more difficult to directly define the switches into the variable.

**Comment by jimav** [comment]
Thank you for the comprehensive response.   I wish other developers paid as much attention as you to  hassles imposed on innocent users by incompatible changes!   Especially because such changes often appear to not be necessary.  In this case I think that making the simple use cases compatible with the informal  "standard" is more important because it will let future users write correct, working scripts the first time using lz4, and not  spend resources debugging and hacking work-arounds to treat lz4 as a special case.   I expect lz4 to enjoy increasingly wide use for decades to come.  Of course, it is your decision. 

Please consider that someone who skims the man page and sees "lz4 supports a command line syntax similar to gzip", might reasonably assume they can substitute lz4 for gzip abd know what to expect.  Almost certainly they will suffer, spending time debugging why disk usage *increased* 50% after lz4'ing all their files (because the originals were not removed), or why their log files filled up with binary garbage (because if stdout is redirected, compressed data is sent to stdout instead of file.lz4).   These problems are absolutely not a fault in the design of lz4, but simply a consequence of being incompatible with gzip.

I suppose yet another family of names could be invented which enable "standard" behaviors,
for example l4zip, unl4zip, l4zcat, etc.   Then "lz4" etc. could be deprecated but left unchanged.
____________

You are right that "lz4 -" is silent.  The only silence problem occurs when a file name is specified, e.g.
       lz4 filename      # prints real-time progress messages to stderr
For gzip compatibilty, by default nothing should be written to stderr unless there is an error.    With -v, a single line to stderr of the form
       filename:  -xx.y% -- replaced with filename.lz4
You might want to enable the real-time progress messages only with a separate option (e.g. --progress).
____________
  lz4 -v filename >log  2>&1
For gzip compat., this should compress to filename.lz4 and write statistics to stderr (nothing to stdout).
____________
Again, a new family of names (l4zip, etc.) could be introduced if it is necessary to preserve the current behaviors.   Thanks for your consideration.

**Comment by Cyan4973** [comment]
> Please consider that someone who skims the man page and sees "lz4 supports a command line syntax similar to gzip", might reasonably assume they can substitute lz4 for gzip abd know what to expect.

That could be first statement to change on short term : while `lz4` has a lot in common with `gzip`, it's not strictly identical. So the man page could 1) state more clearly that it's not exactly identical 2) list the differences 3) provide hints on how to make `lz4` more similar to `gzip` when need be.

**Comment by mkaluza** [comment]
yup, I'd second that request because I've just hit the same wall. Of course you can usually get away by using some sort of script around it, but not always, and it would just be nice and comfy to have drop-in replacement for gzip (just like bzip2, xz and probably others). Of course changing default behavior would be bad because it could break some compatibility, but introducing another binary that behaves _exactly_ like gzip would probably do it. Like lz4g or lz4z - and people wanting gzip behavoir would get exactly that, while the old binary would work the same way (however --rm switch wouldn't hurt the flexibility I think).

If you're interested I could implement at least part of it (unless it's already being done somewhere)

**Comment by Cyan4973** [comment]
I was considering adding a new command `--rm`.
Removing source file would be opt-in, rather than default.
For command line, it's possible to use `alias` to always enable it.
And scripts should use variables, which then can be set to `lz4 --rm`.

**Comment by mkaluza** [comment]
Yes - opt-in for the lz4 binary is completely ok.
I know they "should", but that's not always possible, as well as using aliases - there are automated use cases that don't always expand aliases (usually they don't - every 'exec' bypasses shell expansion). An extreme case is where path is hardcoded and the only thing you could do is to use a symlink, but then the binary needs to behave exactly the same way. Besides when it behaves differently you have to look it up each time to check for any differences - it wastes time...

And yes - I know you can always live without it and if someone doesn't "like" your tool, they don't have to use it... but since it's being done (and it's very good), why not make it more user friendly and flexible (esp that gzip-like behavior is a kind of standard and it's always a surprise if something works differently). It's simply the DRY idea - if you do it once here, nobody ever will have to do it again (the same thing over and over). I'm talking about another binary of course - I have no intention of changing the default behavior of 'lz4'

**Comment by Cyan4973** [comment]
Yes, if you want to create a different cli with different options, for example to be closer to gzip, then it's fine, we can integrate it in lz4 repository within `contrib` folder.

**Comment by t-mat** [comment]
I write basic code structure of `gzip` and `xz` for further investigation in future:

- gzip has flag variable [`keep`](http://git.savannah.gnu.org/cgit/gzip.git/tree/gzip.c?id=3022fdb32287089d2201e715b7beb133a47ff608#n173) for "do not delete input file(s)".
   - If [`keep` doesn't enabled](http://git.savannah.gnu.org/cgit/gzip.git/tree/gzip.c?id=3022fdb32287089d2201e715b7beb133a47ff608#n1020), it [uses](http://git.savannah.gnu.org/cgit/gzip.git/tree/gzip.c?id=3022fdb32287089d2201e715b7beb133a47ff608#n1030) internal function [`xunlink()`](http://git.savannah.gnu.org/cgit/gzip.git/tree/util.c?id=3022fdb32287089d2201e715b7beb133a47ff608#n318) to remove input (source) file(s).
   - `xunlink()` just [uses POSIX standard function `unlink()`](http://git.savannah.gnu.org/cgit/gzip.git/tree/util.c?id=3022fdb32287089d2201e715b7beb133a47ff608#n324) with some tweak.
- xz has flag variable [`opt_keep_original`](http://git.tukaani.org/?p=xz.git;a=blob;f=src/xz/args.c;h=041c80073e6de088effce621ae53dbaf7984af6f;hb=33e803345765854e065bfd58c72ea30ed7f75173#l23) for `-k` (`--keep`) option.
   - Their file I/O function [`io_close_src()`](http://git.tukaani.org/?p=xz.git;a=blob;f=src/xz/file_io.c;h=9bd515ddc894455aff411f1f6783e6160ed51bc3;hb=33e803345765854e065bfd58c72ea30ed7f75173#l656) uses it to [decide](http://git.tukaani.org/?p=xz.git;a=blob;f=src/xz/file_io.c;h=9bd515ddc894455aff411f1f6783e6160ed51bc3;hb=33e803345765854e065bfd58c72ea30ed7f75173#l689) remove src file or not.
   - [`io_unlink()`](http://git.tukaani.org/?p=xz.git;a=blob;f=src/xz/file_io.c;h=9bd515ddc894455aff411f1f6783e6160ed51bc3;hb=33e803345765854e065bfd58c72ea30ed7f75173#l192) is actual function for removing.
       - `io_unlink()` just [uses POSIX standard function `unlink()`](http://git.tukaani.org/?p=xz.git;a=blob;f=src/xz/file_io.c;h=9bd515ddc894455aff411f1f6783e6160ed51bc3;hb=33e803345765854e065bfd58c72ea30ed7f75173#l248) with some tweak for symlink/race conditions.

My thoughts:

- Since we prefer to use standard C function, we may use `remove()` for lz4cli.
- I'm not sure we should implement gzip's and xz's `unlink()` tweaks.  Because these half-baked tweaks are reasons that we should avoid to implement such functionality even though it's handy :disappointed:

**Comment by Cyan4973** [comment]
I fully agree @t-mat

**Comment by mkaluza** [comment]
there's probably some merit behind those "tweaks" and in xz it's even explicitly stated - to avoid race condition (so the problem probably happend at some point), so it's more of a "safety" kind of tweak than just "handy"... Probably it's worth thinking about when implementing --rm functionality.

**Comment by Cyan4973** [comment]
Considering `lz4` portability objectives, a standard C function is preferable to a posix one.
`remove()` will be fine.

**Comment by Cyan4973** [comment]
@inikep added the opt-in `--rm` command in latest `dev` branch update

**Comment by gaia** [comment]
sorry to drop into this years later, and I don't mean to hijack. This discussion AFAIUnderstand alludes to the possibility of recursively piping a directory into lz4 (via tar) and deleting each file in the source as it gets compressed. Is this possible with lz4?

`rsync --remove-source-files` only deletes the source after the entire directory is traversed/compressed (full completion).

now to really get off topic, when will lz4 natively recurse dirs so that tar tricks are not needed? (maybe never, it's working as intended)

**Comment by Cyan4973** [comment]
The `--rm` command deletes source files as they are processed, so gradually.

Compressing and piping multiple files into `stdout` is generally a bad idea, as the result is a direct concatenation of compressed contents, without any metadata such as the original file name. Therefore such a resulting concatenated stream of frames cannot regenerate the original list of files just by itself.

`lz4` is a compressor, not an archiver. It cannot replace `tar`'s capability to store and regenerate directory trees.

**Comment by gaia** [comment]
"Compressing and piping multiple files into stdout is generally a bad idea" you mean piping tar into lz4 is a bad idea?

**Comment by Cyan4973** [comment]
No, I mean piping multiple files into `stdout`.

Like `lz4 -m file1 file2 file3 -c > compressed_and_concatenated.lz4`. 
The resulting file `compressed_and_concatenated.lz4` contains 3 frames concatenated, each with their respective compressed content which is equivalent to `file1.lz4` + `file2.lz4` + `file3.lz4`, but without any additional metadata, so it can't regenerate the files themselves since it doesn't even know their name.

Piping `tar` into `lz4` is a good idea and works great.

**Comment by jimav** [comment]
> This discussion AFAIUnderstand alludes to the possibility of recursively piping a directory into lz4 (via tar) and deleting each file in the source as it gets compressed.

No, actually this bug report is not about any such fancy stuff.  It is simply that the lz4 command is unnecessarily incompatible with gzip in several behaviors.    For example, the simplest (and IMO most common use case) is to just compress a file, or maybe multiple files in-place:

   `gzip foo.txt   # replaces foo.txt with foo.txt.gz`

and lz4 is not a drop-in replacement (e.g. it does not delete the input file by default after creating the compressed version; also it spews compressed bits to stdout instead of creating the .lz4 file if you happen to redirect stdout to a logfile, i.e. very un-unixish behavior).

Too much time has passed to fix any of this, unless an alternate command name is introduced which enables behavior which conforms to the de-facto standard of gzip.

**Comment by Cyan4973** [comment]
I believe the documentation clearly mentions that, in order to get a behavior similar to `gzip`, 
the recommended command is `lz4 -m --rm`.

**Comment by gaia** [comment]
sorry, this might be getting off topic, but i came across this issue searching for how to compress and delete, file by file, recursively.

is it possible or not with lz4+tar? would be good to leave a complete example here. 

PS: if you want me to open another issue pls lmk.

**Comment by Cyan4973** [comment]
The "normal" way to deal with this situation is to `tar` first, and then pass the outcome to be compressed by `lz4` : 
`tar cvf - DIR | lz4 > archive.tar.lz4`

In which case `lz4` only sees `stdin`, so it's unaware of the existence of "files". 
Only `tar` is, so it's up to `tar` to remove original files as they get packed into `stdin`.

`tar` seems to offer a command for that purpose : 
```
--remove-files
remove files after adding them to the archive
```
but it doesn't tell if it removes original files progressively as they are inserted into the archive,
or if they are only removed at the end, once the archive is complete.

I don't know, since I have not tested it directly. But my guess is that the second choice seems safer, because as long as the archive is not complete, something could happen (such as a `Ctrl+C`, or an error in the pipeline by an external process) which would prevent its completeness, resulting in loss of data if original files were deleted before that point.
Also, it's unclear if the command also removes directories or is limited to files only.
